### PR TITLE
[TRITON] Add Attention support to the bench_models benchmarking script

### PR DESCRIPTION
**Pull request by lucas-santos-amd**
## Motivation

- Add standalone Unified Attention Benchmark.
- Add support for MHA, MLA and Unified Attention to the bench_models benchmarking script. 

## Technical Details

New help text of `bench_models.py`:

```
usage: bench_models.py [-h] [--batch_size BATCH_SIZE [BATCH_SIZE ...]] [--seq_len SEQ_LEN [SEQ_LEN ...]] [--TP {1,2,4,8}]
                       [--metric {throughput,bandwidth,time}] [--model MODEL] [--gemm_layout {TN,TT,NN,NT}] [--mha_layout {bshd,thd}]
                       [--output_file OUTPUT_FILE]
 
Model benchmarking tool
 
options:
  -h, --help            show this help message and exit
  --batch_size BATCH_SIZE [BATCH_SIZE ...]
                        Batch size(s) to sweep. Accepts:
                          Single value:            --batch_size 1
                          Multiple values:         --batch_size 1 2 4
                          Range start:stop:step:   --batch_size 1:8:2
                          Combinations of values and ranges are also accepted.
                        Default: 1.
  --seq_len SEQ_LEN [SEQ_LEN ...]
                        Sequence length(s) to sweep. Accepts:
                          Single value:            --seq_len 512
                          Multiple values:         --seq_len 256 512 1024
                          Range start:stop:step:   --seq_len 128:1024:128
                          Combinations of values and ranges are also accepted.
                        For non-attention kernels, M = batch_size x seq_len is passed as M.
                        Default: 4096.
  --TP {1,2,4,8}        Tensor parallel size. Default: 8.
  --metric {throughput,bandwidth,time}
                        Metric to report (throughput=TFLOPS, bandwidth=GB/s, time=ms). Default: throughput. RoPE reports total flops (total floating-point operations) in a separate column (see note in output).MLA benchmark only reports time (ms).
  --model MODEL         model name filter: case-insensitive regex matched against model name (default: all models). e.g. 'llama3' to include only Llama3 family, 'llama|qwen' to include both Llama and Qwen families, '^(?!.*deepseek)' to exclude DeepSeek family.
                        Available models: Llama3 405B, Llama3 70B, Llama3 8B, GPT-OSS 120B, DeepSeek-R1, Llama4 Maverick, Qwen3-235B-A22B.
  --gemm_layout {TN,TT,NN,NT}
                        GEMM layout. Default: TN.
  --mha_layout {bshd,thd}
                        Multi-head attention (MHA) layout (bshd or thd). Default: thd.
  --output_file OUTPUT_FILE
                        Name for the CSV output file. Default: bench_results.
```

Description of output CSV file:

CSV   column | Description | Type | Relevant   kernels
-- | -- | -- | --
Model | LLM   model name | String | All
Kernel | Kernel   name | String | All
batch_size | Input   batch size | Integer | MHA,   MLA
seq_len | Input   sequence length | Integer | RoPE,   MHA, MLA
B | Batched   GEMM batch size | Integer | Batched   GEMM
M | M   dimension | Integer | GEMM,   batched GEMM, RMSNorm, MoE
N | N   dimension | Integer | GEMM,   batched GEMM, RMSNorm
K | K   dimension | Integer | GEMM,   batched GEMM
gemm_layout | GEMM   layout (TN, TT, NN, NT) | String | GEMM,   batched GEMM
hq | Number   of Q heads | Integer | RoPE,   MHA, MLA
hkv | Number   of KV heads | Integer | RoPE,   MHA, MLA
dqk | Head   size of Q and K | Integer | MHA,   MLA
dv | Head   size of V | Integer | MHA,   MLA
rotary_dim | RoPE   rotary dimension | Integer | RoPE
rotate_style | RoPE   rotate style | String | RoPE
mha_layout | Multi-head Attention layout (thd, bshd) | String | MHA
experts | Number   of experts | Integer | MoE
moe_dim1 | First   MoE dimension (i.e. hidden_size) | Integer | MoE
moe_dim2 | Second   MoE dimension (i.e. moe_intermediate_size*2) | Integer | MoE
topk | Number   of experts per token | Integer | MoE
Performance metric | Can   be time in ms, throughput in tflops or bandwidth in gpbs | Decimal | GEMM,   batched GEMM, RMSNorm, MoE, MHA
time(ms) | *Time in ms | Decimal | MLA
total_flops(tflops) | **Total floating-point operations | Decimal | RoPE

*MLA benchmark only reports Time(ms)
**RoPE reports only total floating-point operations, not throughput (TFLOPS).

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py [hunk 1]
```diff
@@ -284,22 +330,177 @@ def build_result_row(self, bench_result: float | str) -> ResultRow:
...
+            s = shape.copy()
+            s["hq"] = max(shape["hq"] // self._tp, 1)
+            s["hkv"] = max(shape["hkv"] // self._tp, 1)
+            result.append(s)
+        return result
+
+    def build_args(self) -> str:
+        shape = self._shape
+        return (
+            f"-mode fwd -causal true --layout bshd --dtype bf16 -b {self._batch_size} "
```
Comment (brunomazzottiamd) [inline op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py]: Please double check if `bshd` layout is the proper one for this benchmark. You can also choose `thd` layout.

File: op_tests/op_benchmarks/triton/model_benchmarking_tool/model_shapes.json [hunk 2]
```diff
@@ -573,6 +608,13 @@
                 "dqk": 128,
                 "dv": 128
             }
+        ],
+        "unified_attention": [
```
Comment (brunomazzottiamd) [inline op_tests/op_benchmarks/triton/model_benchmarking_tool/model_shapes.json]: **Question:** Did you add a unified attention kernel for all models? Is this the right thing to do?

File: op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py [hunk 3]
```diff
@@ -284,22 +330,177 @@ def build_result_row(self, bench_result: float | str) -> ResultRow:
...
+
+
 def _get_handler(kernel: str) -> KernelHandler:
     if "moe" in kernel:
         return MoeKernelHandler()
     if kernel == "rmsnorm":
         return RmsnormKernelHandler()
     if kernel == "rope":
         return RopeKernelHandler()
+    if kernel == "mha":
```
Comment (brunomazzottiamd) [inline op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py]: **Suggestion:**

I think `_get_handler` function is becoming hard to maintain. I don't know if we can do anything better...

Something like this comes to my mind:

```python
_HANDLER_RULES: list[tuple[Callable[[str], bool], type[KernelHandler]]] = [
    (lambda k: "moe" in k, MoeKernelHandler),
    (lambda k: "gemm" in k, GemmKernelHandler),
    (lambda k: k == "rmsnorm", RmsnormKernelHandler),
    (lambda k: k == "rope", RopeKernelHandler),
    (lambda k: k == "mha", MhaKernelHandler),
    (lambda k: k == "mla", MlaKernelHandler),
    (lambda k: k == "unified_attention", UnifiedAttnKernelHandler),
]

def _get_handler(kernel: str) -> KernelHandler:
    for predicate, handler_cls in _HANDLER_RULES:
        if predicate(kernel):
            return handler_cls()
    raise ValueError(f"Kernel {kernel} not supported")
```

Maybe it's more complex than the `if` statements... Please use your own judgement.

File: op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py [hunk 4]
```diff
@@ -284,22 +330,177 @@ def build_result_row(self, bench_result: float | str) -> ResultRow:
...
+
+class UnifiedAttnKernelHandler(KernelHandler):
+    """Handler for unified attention benchmarks (bench_unified_attention.py)."""
+
+    def get_tp_shapes(self, shapes: list[ShapeDict]) -> list[ShapeDict]:
+        result = []
+        for shape in shapes:
+            s = shape.copy()
+            s["hq"] = max(shape["hq"] // self._tp, 1)
+            s["hkv"] = max(shape["hkv"] // self._tp, 1)
```
Comment (brunomazzottiamd) [inline op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py]: **Suggestion:**

This computation is repeated over and over:

```python
s[key] = max(s[key] // self._tp, 1)

return [{**s, "Dim2": max(s["Dim2"] // self._tp, 1)} for s in shapes]

s["num_heads"] = max(s["num_heads"] // self._tp, 1)
s["num_kv_heads"] = max(s["num_kv_heads"] // self._tp, 1)

s["hq"] = max(shape["hq"] // self._tp, 1)
s["hkv"] = max(shape["hkv"] // self._tp, 1)

s["hq"] = max(shape["hq"] // self._tp, 1)
s["hkv"] = max(shape["hkv"] // self._tp, 1)

s["hq"] = max(shape["hq"] // self._tp, 1)
s["hkv"] = max(shape["hkv"] // self._tp, 1)
```

I'd add some utility methods to `KernelHandler` class, to avoid repetition:

```python
def shard(self, value: int) -> int:
    return max(value // self._tp, 1)

def shard_keys(self, s: dict, keys: list[str]) -> None:
    for key in keys:
        s[key] = self._shard(s[key])
```

Usage examples:

```python
return [{**s, "Dim2": self.shard(s["Dim2"])} for s in shapes]

self._shard_keys(s, ["num_heads", "num_kv_heads"])

self._shard_keys(s, ["hq", "hkv"])
```

File: op_tests/op_benchmarks/triton/bench_unified_attention.py [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+
+    max_num_blocks_per_seq = (max_kv_len + block_size - 1) // block_size
+    block_tables = torch.randint(
+        0,
+        num_blocks,
+        (num_seqs, max_num_blocks_per_seq),
+        dtype=torch.int32,
+        device="cuda",
+    )
+    values = torch.arange(0, num_blocks, dtype=torch.int32)
```
Comment (brunomazzottiamd) [inline op_tests/op_benchmarks/triton/bench_unified_attention.py]: **Suggestion:**

Please try to avoid moving data from host to GPU with `.to("cuda")`. I'd try to generate `values` tensor in GPU:

```diff
-    values = torch.arange(0, num_blocks, dtype=torch.int32)
-    values = values[torch.randperm(num_blocks)]
+    values = torch.arange(0, num_blocks, dtype=torch.int32, device="cuda")
+    values = values[torch.randperm(num_blocks, device="cuda")]
     block_tables = (
         values[: num_seqs * max_num_blocks_per_seq]
         .view(num_seqs, max_num_blocks_per_seq)
         .contiguous()
-        .to("cuda")
     )
```

File: op_tests/op_benchmarks/triton/bench_unified_attention.py [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+    value_cache = torch.randn_like(key_cache)
+    cu_query_lens = torch.tensor(
+        [0] + query_lens,
+        dtype=torch.int32,
+        device="cuda",
+    ).cumsum(dim=0, dtype=torch.int32)
+    kv_lens = torch.tensor(kv_lens, dtype=torch.int32, device="cuda")
+
+    max_num_blocks_per_seq = (max_kv_len + block_size - 1) // block_size
+    block_tables = torch.randint(
```
Comment (brunomazzottiamd) [inline op_tests/op_benchmarks/triton/bench_unified_attention.py]: I think this code can be removed:

```python
    block_tables = torch.randint(
        0,
        num_blocks,
        (num_seqs, max_num_blocks_per_seq),
        dtype=torch.int32,
        device="cuda",
    )
```

`block_tables` is immediately overwritten, and the data generated by this statement never used as far as I can see.

File: op_tests/op_benchmarks/triton/bench_unified_attention.py [hunk 7]
```diff
@@ -0,0 +1,578 @@
+import math
```
Comment (brunomazzottiamd) [inline op_tests/op_benchmarks/triton/bench_unified_attention.py]: I did a quick review of `op_tests/op_benchmarks/triton/bench_unified_attention.py` and couldn't see anything seriously wrong with it. However, a thumbs up from @cagrikymk is mandatory IMHO (he doesn't need to check the entire PR, just `bench_unified_attention.py`).

**Review by brunomazzottiamd — DISMISSED** [review]
@lucas-santos-amd, I have some suggestions, but no blockers. It would be nice to have a thumbs up from Cagri regarding `op_tests/op_benchmarks/triton/bench_unified_attention.py`.

File: op_tests/op_benchmarks/triton/model_benchmarking_tool/model_shapes.json (hunk 2 above)
Comment (lucas-santos-amd) [reply op_tests/op_benchmarks/triton/model_benchmarking_tool/model_shapes.json]: All of the current available ones, except deepseek-r1

File: op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py (hunk 1 above)
Comment (lucas-santos-amd) [reply op_tests/op_benchmarks/triton/model_benchmarking_tool/bench_models.py]: Looks like we want both, so I added a CLI argument for the MHA layout

File: op_tests/op_benchmarks/triton/bench_unified_attention.py (hunk 6 above)
Comment (lucas-santos-amd) [reply op_tests/op_benchmarks/triton/bench_unified_attention.py]: Removed unified attention from this branch

File: op_tests/op_benchmarks/triton/bench_unified_attention.py (hunk 5 above)
Comment (lucas-santos-amd) [reply op_tests/op_benchmarks/triton/bench_unified_attention.py]: Removed unified attention from this branch

File: op_tests/op_benchmarks/triton/bench_unified_attention.py (hunk 7 above)
Comment (lucas-santos-amd) [reply op_tests/op_benchmarks/triton/bench_unified_attention.py]: Removed unified attention from this branch

**Comment by lucas-santos-amd** [comment]
Support for unified attention has been removed from this branch. It will be added back in a later PR once https://github.com/ROCm/aiter/pull/2190 is merged.
